src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/Login/FeatureExtensions.cs

```diff
@@ -0,0 +1,631 @@
+                await WriteIdentifierAsync(context, reconnectData._language != reconnectData._initialLanguage ? reconnectData._language : null, isAsync, ct).ConfigureAwait(false);
+                for (int i = 0; i < SessionData._maxNumberOfSessionStates; i++)
+                {
+                    if (writeState[i])
+                    {
+                        await writer.WriteByteAsync((byte)i, isAsync, ct).ConfigureAwait(false);
+                        if (reconnectData._delta[i]._dataLength < 0xFF)
+                        {
+                            await writer.WriteByteAsync((byte)reconnectData._delta[i]._dataLength, isAsync, ct).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            await writer.WriteByteAsync(0xFF, isAsync, ct).ConfigureAwait(false);
+                            await writer.WriteIntAsync(reconnectData._delta[i]._dataLength, isAsync, ct).ConfigureAwait(false);
+                        }
+                        await writer.WriteBytesAsync(reconnectData._delta[i]._data.AsMemory(0, reconnectData._delta[i]._dataLength), isAsync, ct).ConfigureAwait(false);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Whether to request session recovery.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public static bool ShouldRequest(LoginHandlerContext context) => context.ConnectionOptions.ConnectRetryCount > 0;
+
+        private async ValueTask WriteCollationAsync(LoginHandlerContext context, SqlCollation collation, bool isAsync, CancellationToken ct)
+        {
+            TdsStream stream = context.TdsStream;
+
+            if (collation == null)
+            {
+                await stream.TdsWriter.WriteByteAsync(0, isAsync, ct).ConfigureAwait(false);
+            }
+            else
+            {
+                await stream.TdsWriter.WriteByteAsync(sizeof(uint) + sizeof(byte), isAsync, ct).ConfigureAwait(false);
+                await stream.TdsWriter.WriteUnsignedIntAsync(collation._info, isAsync, ct).ConfigureAwait(false);
+                await stream.TdsWriter.WriteByteAsync(collation._sortId, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        private static async ValueTask WriteIdentifierAsync(LoginHandlerContext context, string s, bool isAsync, CancellationToken ct)
+        {
+            TdsStream stream = context.TdsStream;
+            if (null != s)
+            {
+                await stream.TdsWriter.WriteByteAsync(checked((byte)s.Length), isAsync, ct).ConfigureAwait(false);
+                await stream.WriteStringAsync(s, isAsync, ct).ConfigureAwait(false);
+            }
+            else
+            {
+                await stream.WriteByteAsync((byte)0, isAsync, ct).ConfigureAwait(false);
```

Comment:
A few places still need TdsWriter to be used, not critical but something to address maybe later :)